### DISCUSSION TITLE
Detach causes servo to change position with multiple servos

### DISCUSSION THREAD

**Discussion by darkgrue** [ArminJo/ServoEasing#91] [Q&A]
I've noticed a problem with my two-servo implementation. I detach the servos after performing a series of moves to take loads off the servos, etc. I've noticed the second servo will sometimes (not consistently, but often) change position rapidly on the detach.

```
Serial.println("Detaching Servo 1...");
Servo1.detach();
Serial.println("Detaching Servo 2...");
Servo2.detach();
```

With `TRACE` defined, I don't see any output between the detach statements, so nothing is happening that's logged.

```
Detaching Servo 1...
Detaching Servo 2...
```

I made a rough guess that maybe the current PWM value is getting altered, because I read this in the docs:

> If you detach a servo and then attach another one, the latter will get the index of the former detached one.

and thought maybe that shift was changing the servo position as some sort of side-effect. It's probably a bad guess, but by writing the current position of the second servo again before detaching it, it seems to have worked around the problem:

```
  Serial.println("Detaching Servo 1...");
  Servo1.detach();
  Servo2.setEaseTo(S2AngleLast);
  Serial.println("Detaching Servo 2...");
  Servo2.detach();
```

```
Detaching Servo 1...
1/33: 115|1729 -> 115|1729 = 0|0 in 0 ms with speed=5 and easingType=0x0|linear_in MillisAtStartMove=1803
Detaching Servo 2...
```

Is there some sort of bug going on in the library that causes this? None of the servos should move when any other servo is detached (including itself), correct? These are dumb servos, they don't have a signal loss position or anything like that.

**Comment by i3creations** [comment]
I ran into a similar issue.  Instead of `detach()` I called `release()` instead and it appears to have resolved my issue.  `release()` appears to halt sending any PWM signals to the servo(s).  I am using an ESP32-C3 module in this instance.  However, when using an Arduino Nano ATMega328P module the `detach()` function does not alter the position of the servo.

Hope this helps!

**Reply by darkgrue** [reply]
I'm not seeing a release() function associated with any of the libraries in the chain (easing or ESP32Servo), so I'm curious as to how this accomplishes the goal? I gave it a try, and servo moves stopped working after the first time they're release()'d.

Also, what do you do about all the array handling that the easing and ESP32Servo are doing in detach()? Are you decrementing those arrays yourself, or just ignoring them?

**Comment by darkgrue** [comment]
I just realized something. Maybe I'm using the _wrong_ ESP32 servo library? I bet there's more than one...

Which one is this developed with? I've been using https://github.com/madhephaestus/ESP32Servo

**Reply by ArminJo** [reply]
This is the right one, but it seems, that it may have problems with detach. I am trying to reproduce this problem, but it may take a few days.

**Reply by darkgrue** [reply]
It's very inconsistent. I've had times when it will do it over and over, and other times when it won't happen at all. Which makes it really hard to reproduce, let alone solve...

Appreciate the help though, this problem has been vexing us for a while, and I just can't figure out what's going on myself.

**Reply by darkgrue** [reply]
It looks like a `release()` method was added in v1.12 of ESP32Servo to accomplish what I was trying to achieve with attaching and detaching between moves (but does so with a lot less overhead). Since it's a _very_ difficult bug to reproduce, I'm not sure it's solved it, but it does seem to do what I wanted. I'm giving it a try and watching to see if the bug reoccurs.

**Reply by ArminJo** [reply]
I opened an [issue at ESP32Servo](https://github.com/madhephaestus/ESP32Servo/issues/46) for it.
